### PR TITLE
arm64: decrustify and extend SMP boot code

### PR DESCRIPTION
**Pull request by npitre**

### CODE REVIEW

**Comment by npitre** [comment]
This supersedes PR #34271 with additional code refactoring.

@JiafeiPan please review.

File: arch/arm64/core/smp.c [hunk 1]
```diff
@@ -22,53 +23,103 @@
...
-	arm64_cpu_init[cpu_num].fn = fn;
-	arm64_cpu_init[cpu_num].arg = arg;
-	arm64_cpu_init[cpu_num].sp =
-		(void *)(Z_THREAD_STACK_BUFFER(stack) + sz);
+	dsb();
 
-	arch_dcache_range((void *)&arm64_cpu_init[cpu_num],
-			  sizeof(arm64_cpu_init[cpu_num]), K_CACHE_WB_INVD);
+	/* store mpid last as this is our synchronization point */
+	arm64_cpu_boot_params.mpid = cpu_mpid;
```
Comment (JiafeiPan) [inline arch/arm64/core/smp.c:91]: move dsb() from ahead of this line to behind of this line.

File: arch/arm64/core/smp.c [hunk 2]
```diff
@@ -22,53 +23,103 @@
...
 
-volatile struct {
-	void *sp; /* Fixed at the first entry */
+struct boot_params {
+	uint64_t mpid;
+	char *sp;
 	arch_cpustart_t fn;
 	void *arg;
-} __aligned(L1_CACHE_BYTES) arm64_cpu_init[CONFIG_MP_NUM_CPUS];
+	int cpu_num;
```
Comment (JiafeiPan) [inline arch/arm64/core/smp.c:36]: Don't need to reserve cpu_num anymoe because we use single boot_params for all secondary cores.

**Review by JiafeiPan — COMMENTED** [review]
> This supersedes PR #34271 with additional code refactoring.
> 
> @JiafeiPan please review.

Hi, npitre,  Thanks for your refined my patch, but could you please add my signed-off in the second patch, thanks.

**Comment by npitre** [comment]
> Hi, npitre,  Thanks for your refined my patch, but could you please add my
> signed-off in the secondary patch, thanks.

Sure.

> +     /* store mpid last as this is our synchronization point */
> +     arm64_cpu_boot_params.mpid = cpu_mpid;
>
> move dsb() from ahead of this line to behind of this line.

No, it has to be between setting mpid and the rest to guarantee that all
the structure updates have been sent to memory first and that the mpid
update will be seen last. Otherwise the secondary could see its MPID and load
its stack pointer before it was sent to memory.

The `arch_dcache_range()` should act as implicit data synchronization
already.

>       void *arg;
> -} __aligned(L1_CACHE_BYTES) arm64_cpu_init[CONFIG_MP_NUM_CPUS];
> +     int cpu_num;
>
> Don't need to reserve cpu_num anymoe because we use single boot_params for
> all secondary cores.

Secondary cores still need to know what their assigned logical CPU
number is.

**Comment by JiafeiPan** [comment]
> > Hi, npitre,  Thanks for your refined my patch, but could you please add my
> > signed-off in the secondary patch, thanks.
> 
> Sure.
> 
> > * ```
> >   /* store mpid last as this is our synchronization point */
> >   ```
> > * ```
> >   arm64_cpu_boot_params.mpid = cpu_mpid;
> >   ```
> > 
> > move dsb() from ahead of this line to behind of this line.
> 
> No, it has to be between setting mpid and the rest to guarantee that all
> the structure updates have been sent to memory first and that the mpid
> update will be seen last. Otherwise the secondary could see its MPID and load
> its stack pointer before it was sent to memory.

Per my understand, only when pm_cpu_on() is called, secondary can only boot into Zephyr, so currently in this position, secondary is still in WFE or WFI status.

> 
> The `arch_dcache_range()` should act as implicit data synchronization
> already.
> 
> > ```
> >   void *arg;
> > ```
> > 
> > 
> > -} __aligned(L1_CACHE_BYTES) arm64_cpu_init[CONFIG_MP_NUM_CPUS];
> > 
> > * ```
> >   int cpu_num;
> >   ```
> > 
> > Don't need to reserve cpu_num anymoe because we use single boot_params for
> > all secondary cores.
> 
> Secondary cores still need to know what their assigned logical CPU
> number is.

Yes, I see.

**Comment by npitre** [comment]
> > No, it has to be between setting mpid and the rest to guarantee that all
> > the structure updates have been sent to memory first and that the mpid
> > update will be seen last. Otherwise the secondary could see its MPID and load
> > its stack pointer before it was sent to memory.

> Per my understand, only when pm_cpu_on() is called, secondary can only boot into Zephyr, so currently in this position, secondary is still in WFE or WFI status.

In that case it makes no difference. But if for some reason all 
secondary CPUs were released together then we still have our own gating 
mechanism to ensure an orderly bringup.

**Comment by JiafeiPan** [comment]
> > No, it has to be between setting mpid and the rest to guarantee that all > the structure updates have been sent to memory first and that the mpid > update will be seen last. Otherwise the secondary could see its MPID and load > its stack pointer before it was sent to memory. Per my understand, only when pm_cpu_on() is called, secondary can only boot into Zephyr, so currently in this position, secondary is still in WFE or WFI status.
> In that case it makes no difference. But if for some reason all secondary CPUs were released together then we still have our own gating mechanism to ensure an orderly bringup.

Current design only support to boot secondary one after another because we use the same arm64_cpu_boot_params instance for all secondary. If we need to support the case that all secondary CPUs will be released together, we need to revert back to use different arm64_cpu_boot_params instance for different secondary core.

**Comment by JiafeiPan** [comment]
> > No, it has to be between setting mpid and the rest to guarantee that all > the structure updates have been sent to memory first and that the mpid > update will be seen last. Otherwise the secondary could see its MPID and load > its stack pointer before it was sent to memory. Per my understand, only when pm_cpu_on() is called, secondary can only boot into Zephyr, so currently in this position, secondary is still in WFE or WFI status.
> In that case it makes no difference. But if for some reason all secondary CPUs were released together then we still have our own gating mechanism to ensure an orderly bringup.

Current design only support to boot secondary one after another because we use the same arm64_cpu_boot_params instance for all secondary. If we need to support the case that all secondary CPUs will be released together, we need to revert back to use different arm64_cpu_boot_params instance for different secondary core.

**Comment by npitre** [comment]
On Sun, 18 Apr 2021, Jiafei Pan wrote:

> > > No, it has to be between setting mpid and the rest to guarantee that all > the structure updates have been sent to memory first and that the mpid > update will be seen last. Otherwise the secondary could see its MPID and load > its stack pointer before it was sent to memory. Per my understand, only when pm_cpu_on() is called, secondary can only boot into Zephyr, so currently in this position, secondary is still in WFE or WFI status.
> > In that case it makes no difference. But if for some reason all secondary CPUs were released together then we still have our own gating mechanism to ensure an orderly bringup.
> 
> Current design only support to boot secondary one after another 
> because we use the same arm64_cpu_boot_params instance for all 
> secondary.

Exact.

> If we need to support the case that all secondary CPUs will be 
> released together, we need to revert back to use different 
> arm64_cpu_boot_params instance for different secondary core.

No! That's the whole point of having a synchronization mechanism using 
the MPID value to let each CPU enter `z_arm64_secondary_prep_c` in an 
orderly manner.

Please look at reset.S starting at line 107 again.

**Comment by JiafeiPan** [comment]
ensure an orderly bringup. Current design only support to boot secondary one after another because we use the same arm64_cpu_boot_params instance for all secondary.
> Exact.
> If we need to support the case that all secondary CPUs will be released together, we need to revert back to use different arm64_cpu_boot_params instance for different secondary core.
> No! That's the whole point of having a synchronization mechanism using the MPID value to let each CPU enter `z_arm64_secondary_prep_c` in an orderly manner. Please look at reset.S starting at line 107 again.

Yes, I see, my point is currently I don't understand why we need dsb here, maybe we need it in future design. but anyway no other impaction if we keep it.

**Review by carlocaione — APPROVED** [review]

**Review by povergoing — APPROVED** [review]

**Pull request merged**